packages/protocol/src/action-protocol/base-protocol.ts

```diff
@@ -13,114 +13,224 @@
+    /**
+     * A human-readable description of the reject reason. Typically this is an error message
+     * that has been thrown when handling the corresponding {@link RequestAction}.
+     */
+    message: string;
+    /**
+     * Optional additional details.
+     */
+    detail?: JsonPrimitive;
 }
 
-export function isRejectAction(action: any): action is RejectAction {
-    return isResponseAction(action) && action.kind === RejectAction.KIND && isString(action, 'message');
+export namespace RejectAction {
+    export const KIND = 'rejectRequest';
+
+    export function is(object: any): object is RejectAction {
+        return Action.hasKind(object, RejectAction.KIND) && hasStringProp(object, 'message');
+    }
+
+    export function create(message: string, options: { detail?: JsonPrimitive; responseId?: string } = {}): RejectAction {
+        return {
+            kind: KIND,
+            responseId: '',
+            message,
+            ...options
+        };
+    }
 }
 
 /**
  * Operations are actions that denote requests from the client to _modify_ the model. Model modifications are always performed by the
  * server. After a successful modification, the server sends the updated model back to the client using the `UpdateModelAction`.
+ * An operation contains a dedicated `isOperation` property that is used as a discriminator. This is necessary so that the server
+ * can distinguish between plain actions and operations.
+ * The corresponding namespace offers a helper function for type guard checks.
  */
-export interface Operation extends Action {}
+export interface Operation extends Action {
+    /**
+     * Discriminator property to make operations distinguishable from plain {@link Action}s.
+     */
+    isOperation: true;
+}
+
+export namespace Operation {
+    export function is(object: any): object is Operation {
+        return Action.is(object) && (object as any).isOperation === true;
+    }
+
+    /**
+     * Typeguard function to check wether the given object is an {@link Operation} with the given `kind`.
+     * @param object The object to check.
+     * @param kind  The expected operation kind.
+     * @returns A type literal indicating wether the given object is an operation with the given kind.
+     */
+    export function hasKind(object: any, kind: string): object is Operation {
+        return Operation.is(object) && object.kind === kind;
+    }
+}
 
 /**
  * An operation that executes a list of sub-operations.
+ * The corresponding namespace declares the action kind as constant and offers helper functions for type guard checks
+ * and creating new `CompoundOperations`.
  */
-export class CompoundOperation implements Operation {
-    static readonly KIND = 'compound';
-    constructor(public operationList: Operation[], public readonly kind: string = CompoundOperation.KIND) {}
+export interface CompoundOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof CompoundOperation.KIND;
```

Comment:
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.